File: app/src/test/java/com/coinninja/coinkeeper/view/button/ConfirmHoldButtonTest.java
package com.coinninja.coinkeeper.view.button;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.Animation;

import com.coinninja.coinkeeper.R;
import com.coinninja.coinkeeper.TestCoinKeeperApplication;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
@Config(application = TestCoinKeeperApplication.class)
public class ConfirmHoldButtonTest {

    private ConfirmHoldButton confirmHoldButton;
    private MotionEvent motionEventDown;
    private MotionEvent motionEventUp;
    private ObjectAnimator progressRingAnimation;
    private Animation growAnimation;
    private ConfirmHoldButton.OnConfirmHoldEndListener onConfirmHoldEndListener;

    @Before
    public void setUp() throws Exception {
        Context context = RuntimeEnvironment.application.getApplicationContext();
        View layout = LayoutInflater.from(context).inflate(R.layout.fragment_confirm_pay_dialog, null, false);

        confirmHoldButton = layout.findViewById(R.id.confirm_pay_hold_progress_btn);
        motionEventDown = mock(MotionEvent.class);
        motionEventUp = mock(MotionEvent.class);

        when(motionEventDown.getAction()).thenReturn(MotionEvent.ACTION_DOWN);
        when(motionEventUp.getAction()).thenReturn(MotionEvent.ACTION_UP);

        onConfirmHoldEndListener = mock(ConfirmHoldButton.OnConfirmHoldEndListener.class);
        progressRingAnimation = mock(ObjectAnimator.class);
        growAnimation = mock(Animation.class);

        confirmHoldButton.progressRingAnimation(progressRingAnimation);
        confirmHoldButton.setGrowAnimation(growAnimation);
        confirmHoldButton.setOnConfirmHoldEndListener(onConfirmHoldEndListener);
    }

    @Test
    public void touch_and_hold() {

        confirmHoldButton.onTouch(confirmHoldButton, motionEventDown);
        confirmHoldButton.getCountDownRunnable().run();

        verify(progressRingAnimation).start();
        verify(growAnimation).start();
        verify(onConfirmHoldEndListener).onHoldCompleteSuccessfully();
    }

    @Test
    public void touch_lift_finger() {

        confirmHoldButton.onTouch(confirmHoldButton, motionEventUp);

        verify(progressRingAnimation).cancel();
        verify(growAnimation).cancel();
    }
}

File: app/src/test/java/com/coinninja/coinkeeper/view/button/ConfirmHoldButtonTest.kt
package com.coinninja.coinkeeper.view.button

import android.animation.ObjectAnimator
import android.view.MotionEvent
import android.view.animation.Animation
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.coinninja.coinkeeper.R
import com.coinninja.coinkeeper.ui.base.TestableActivity
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric


@RunWith(AndroidJUnit4::class)
class ConfirmHoldButtonTest {

    private lateinit var confirmHoldButton: ConfirmHoldButton
    private val motionEventDown: MotionEvent = mock()
    private val motionEventUp: MotionEvent = mock()
    private val progressRingAnimation: ObjectAnimator = mock()
    private val growAnimation: Animation = mock()
    private val onConfirmHoldEndListener: ConfirmHoldButton.OnConfirmHoldEndListener = mock()

    @Before
    fun setUp() {
        val activity = Robolectric.setupActivity(TestableActivity::class.java)
        activity.appendLayout(R.layout.activity_confirm_payment)
        confirmHoldButton = activity.findViewById(R.id.confirm_pay_hold_progress_btn)

        whenever(motionEventDown.action).thenReturn(MotionEvent.ACTION_DOWN)
        whenever(motionEventUp.action).thenReturn(MotionEvent.ACTION_UP)

        confirmHoldButton.progressRingAnimation(progressRingAnimation)
        confirmHoldButton.setGrowAnimation(growAnimation)
        confirmHoldButton.setOnConfirmHoldEndListener(onConfirmHoldEndListener)
    }

    @Test
    fun touch_and_hold() {

        confirmHoldButton.onTouch(confirmHoldButton, motionEventDown)
        confirmHoldButton.countDownRunnable.run()

        verify(progressRingAnimation).start()
        verify(growAnimation).start()
        verify(onConfirmHoldEndListener).onHoldCompleteSuccessfully()
    }

    @Test
    fun touch_lift_finger() {

        confirmHoldButton.onTouch(confirmHoldButton, motionEventUp)

        verify(progressRingAnimation).cancel()
        verify(growAnimation).cancel()
    }
}